BE: reset dynamo before each test in test_torch.py

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131397
* #131372
* __->__ #131388
* #131551

https://github.com/pytorch/pytorch/pull/126586 tried to reset dynamo before each unit test. That PR get reverted a couple of times because we see post-land test failures that we don't see before merge. This PR only reset dynamo before each tests in `test_torch.py` to make it easier to land.

Eventually after we reset dynamo in each individual test files, we can move the change to the base class (TestCase) and remove the change in individual test files.

shunting314 committed:
BE: reset dynamo before each test in test_torch.py

[ghstack-poisoned]

zou3519 commented on test/test_torch.py:
```diff
@@ -140,6 +140,13 @@ def _rand_shape(self, dim, min_size, max_size):
 
+    def setUp(self):
+        TestCase.setUp(self)
+        # reset dynamo cache to avoid issues like
+        # https://github.com/pytorch/pytorch/issues/125967#issuecomment-2118483919
+        # which depends on test order.
+        torch._dynamo.reset()
```
torch._dynamo.reset() might be expensive (when called so many times), so you might want to guard this on if we are doing PYTORCH_TEST_WITH_INDUCTOR

zou3519 replied:
Actually, see my next point

zou3519 commented on test/test_torch.py:
```diff
@@ -6428,6 +6435,13 @@ def test_masked_scatter_inplace_noncontiguous(self, device):
 class TestDevicePrecision(TestCase):
     exact_dtype = True
 
+    def setUp(self):
+        TestCase.setUp(self)
+        # reset dynamo cache to avoid issues like
+        # https://github.com/pytorch/pytorch/issues/125967#issuecomment-2118483919
+        # which depends on test order.
+        torch._dynamo.reset()
+
```
We probably want to do something like delete lines in the following: https://www.internalfb.com/code/fbsource/[8fc78181ceb55af1103ed4145ea1510d5f6335bb]/xplat/caffe2/torch/testing/_internal/dynamo_test_failures.py?lines=66-70. 

This would save us the trouble of needing to add .reset() everytime someone adds a new TestCase.

Also, the above list turns suppress_errors to False, you might want a separate list for that because that is more work.

shunting314 replied:
So I can add a list like:
```
FIXME_inductor_dont_reset_dynamo = {
    "test_modules",
    "test_ops",
    "test_ops_gradients",
    "test_torch",
}
```
And land this. With follow up PRs, we remove items from the list. Once the list is empty, we remove it completely?

zou3519 replied:
Yes

shunting314 replied:
Make the change in a base PR and convert this PR to a one line change

shunting314 committed:
Update on "BE: reset dynamo before each test in test_torch.py"


https://github.com/pytorch/pytorch/pull/126586 tried to reset dynamo before each unit test. That PR get reverted a couple of times because we see post-land test failures that we don't see before merge. This PR only reset dynamo before each tests in `test_torch.py` to make it easier to land.

Eventually after we reset dynamo in each individual test files, we can move the change to the base class (TestCase) and remove the change in individual test files.



[ghstack-poisoned]

shunting314 committed:
Update on "BE: reset dynamo before each test in test_torch.py"


https://github.com/pytorch/pytorch/pull/126586 tried to reset dynamo before each unit test. That PR get reverted a couple of times because we see post-land test failures that we don't see before merge. This PR only reset dynamo before each tests in `test_torch.py` to make it easier to land.

Eventually after we reset dynamo in each individual test files, we can move the change to the base class (TestCase) and remove the change in individual test files.



[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge

shunting314 committed:
Update on "BE: reset dynamo before each test in test_torch.py"


https://github.com/pytorch/pytorch/pull/126586 tried to reset dynamo before each unit test. That PR get reverted a couple of times because we see post-land test failures that we don't see before merge. This PR only reset dynamo before each tests in `test_torch.py` to make it easier to land.

Eventually after we reset dynamo in each individual test files, we can move the change to the base class (TestCase) and remove the change in individual test files.



[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge -i

The failed check is due to timeout
